Rewrite: CSS namespace module and add CSS & HTML to glossary (#27664)
mdn/content: files/en-us/web/css/css_namespaces/index.md

{{CSSRef}}

The **CSS namespaces** module defines the syntax for using {{glossary("namespace", "namespaces")}} in CSS.

> CSS isn't just for HTML. A single stylesheet may be used to style SVG, MathML, XML, HTML, each of which has a different namespace, or a document containing multiple namespaces. Element tag names are not all unique to a single language. For example, The `<a>` element isn't limited to HTML. You may want to use different CSS style for your SVG differently than for your HTML links, or ensure that {{domxref("Document.querySelectorAll", "querySelectorAll(\"a\")")}} selects the right kind of element. [The `@namespace` rule](/en-US/docs/Web/CSS/@namespace), defined in this module, enable distinguishing between same-named elements in different namespaces.

Reviewer: this needs rephrasing "You may want to use different CSS style for your SVG differently than for your HTML links"

```suggestion
CSS isn't just for styling HTML. A stylesheet may be used to style SVG, MathML, XML, or HTML, each of which has a different namespace or a document containing multiple namespaces. Element tag names are not unique to a single language. For example, the `<a>` element isn't limited to HTML. You may want to use different CSS style for your SVG differently than for your HTML links or ensure that {{domxref("Document.querySelectorAll", "querySelectorAll(\"a\")")}} selects the right kind of element. The [`@namespace`](/en-US/docs/Web/CSS/@namespace) at-rule defined in this module enables distinguishing between same-named elements in different namespaces.
```

> CSS isn't just for styling HTML. A stylesheet may be used to style SVG, MathML, XML, or HTML, each of which has a different namespace or a document containing multiple namespaces. Element tag names are not unique to a single language. For example, the `<a>` element isn't limited to HTML. You may want to use different CSS style for your SVG differently than for your HTML links or ensure that {{domxref("Document.querySelectorAll", "querySelectorAll(\"a\")")}} selects the right kind of element. The [`@namespace`](/en-US/docs/Web/CSS/@namespace) at-rule defined in this module enables distinguishing between same-named elements in different namespaces.

Author: ```suggestion
CSS isn't just for styling HTML. A stylesheet may be used to style SVG, MathML, XML, or HTML, each of which has a different namespace or a document containing multiple namespaces. 

The [`@namespace`](/en-US/docs/Web/CSS/@namespace) at-rule defined in this module enables distinguishing between same-named elements in different namespaces. Element tag names are not unique to a single language. For example, the `<a>` element isn't limited to HTML. You may want to style the `<a>`s  within your SVGs differently than links in your HTML. You also likely want to ensure that {{domxref("Document.querySelectorAll", "querySelectorAll(\"a\")")}} selects the right kind of element. Namespacing can help.
```

> The `@namespace` rule is used for declaring a default namespace and for binding namespaces to namespace prefixes. The namespaces module also defines the syntax for using these prefixes to represent namespace-qualified names. That's it. What a name means, or if the name is even valid, depends on the context and host language.

Reviewer: ```suggestion
The `@namespace` rule is used for declaring a default namespace and for binding namespaces to namespace prefixes. The namespaces module also defines the syntax for using these prefixes to represent namespace-qualified names. That's it. What a name means or if the name is even valid depends on the context and host language.
```

## Reference

### At-rules

- {{cssxref("@namespace")}}

## Guides

- [Namespaces crash course](/en-US/docs/Web/SVG/Namespaces_Crash_Course)

  - : Deep dive into what a namespace is and how they are used in XML and XML-based markup languages.

## Related concepts

- CSS [namespace separator (`|`)](/en-US/docs/Web/CSS/Namespace_separator) combinator

> - [CSS type selector](/en-US/docs/Web/CSS/Type_selectors)
> - [CSS universal selector](/en-US/docs/Web/CSS/Universal_selectors)

Reviewer: ```suggestion
- CSS [type selectors](/en-US/docs/Web/CSS/Type_selectors)
- CSS [universal selector](/en-US/docs/Web/CSS/Universal_selectors)
```

- [`Document.createElementNS()`](/en-US/docs/Web/API/Document/createElementNS) method
- [`Document.getElementsByTagNameNS()`](/en-US/docs/Web/API/Document/getElementsByTagNameNS) method
- [`Element.getAttributeNodeNS()`](/en-US/docs/Web/API/Element/getAttributeNodeNS) method
- [`Element.getAttributeNS()`](/en-US/docs/Web/API/Element/getAttributeNS) method
- [`Element.getElementsByTagNameNS()`](/en-US/docs/Web/API/Element/getElementsByTagNameNS) method
- [`Element.hasAttributeNS()`](/en-US/docs/Web/API/Element/hasAttributeNS) method
- [`Element.namespaceURI`](/en-US/docs/Web/API/Element/namespaceURI) property
- [`Element.removeAttributeNS()`](/en-US/docs/Web/API/Element/removeAttributeNS) method
- [`Element.setAttributeNS()`](/en-US/docs/Web/API/Element/setAttributeNS) method
- [`Element.setAttributeNodeNS()`](/en-US/docs/Web/API/Element/setAttributeNodeNS) method
- [`NamedNodeMap.getNamedItemNS()`](/en-US/docs/Web/API/NamedNodeMap/getNamedItemNS) method
- [`NamedNodeMap.removeNamedItemNS()`](/en-US/docs/Web/API/NamedNodeMap/removeNamedItemNS) method
- [`NamedNodeMap.setNamedItemNS()`](/en-US/docs/Web/API/NamedNodeMap/setNamedItemNS) method

> - {{glossary("namespace")}} glossary term

Reviewer: ```suggestion
- {{glossary("Namespace")}} glossary term
```

## Specifications

{{Specifications}}

## Browser compatibility

{{Compat}}

## See also

> - [CSS Namespace example in SVG](/en-US/docs/Web/SVG/Element/a#example)
> - [CSS `url()` functions](/en-US/docs/Web/CSS/url)
> - [CSS at-rules](/en-US/docs/Web/CSS/At-rule)
> - [CSS at-rule functions](/en-US/docs/Web/CSS/At-rule-functions)
> - [CSS selector module](/en-US/docs/Web/CSS/CSS_selectors)

Reviewer: I realize that this is not defined in the "See also" guidelines, but in this case should we add a lead-in for this list, something like "These are a few examples of different namespaces:"?

```suggestion
- [`<a>`](/en-US/docs/Web/SVG/Element/a#example) SVG element
- [CSS `url()` function](/en-US/docs/Web/CSS/url)
- [CSS at-rules](/en-US/docs/Web/CSS/At-rule)
- [CSS at-rule functions](/en-US/docs/Web/CSS/At-rule-functions)
- [CSS selectors](/en-US/docs/Web/CSS/CSS_selectors)
```

{{CSSRef}}

The **CSS namespaces** module defines the syntax for using {{glossary("namespace", "namespaces")}} in CSS.

> CSS isn't just for styling HTML. A stylesheet may be used to style SVG, MathML, XML, or HTML, each of which has a different namespace or a document containing multiple namespaces. 

Reviewer: ```suggestion
CSS isn't just for styling HTML. A stylesheet may be used to style SVG, MathML, XML, or HTML, each of which has a different namespace or a document containing multiple namespaces.
```

> The [`@namespace`](/en-US/docs/Web/CSS/@namespace) at-rule defined in this module enables distinguishing between same-named elements in different namespaces. Element tag names are not unique to a single language. For example, the `<a>` element isn't limited to HTML. You may want to style the `<a>`s within your SVGs differently than links in your HTML. You also likely want to ensure that {{domxref("Document.querySelectorAll", "querySelectorAll(\"a\")")}} selects the right kind of element. Namespacing can help.

Reviewer: ```suggestion
The [`@namespace`](/en-US/docs/Web/CSS/@namespace) at-rule defined in this module enables distinguishing between same-named elements in different namespaces. Element tag names are not unique to a single language. For example, the `<a>` element isn't limited to HTML. You may want to style the `<a>`s within your SVGs differently from the links in your HTML. You also will likely want to ensure that {{domxref("Document.querySelectorAll", "querySelectorAll(\"a\")")}} selects the right kind of element. Namespacing can help.
```

> CSS isn't just for styling HTML. A stylesheet may be used to style SVG, MathML, XML, or HTML, each of which has a different namespace or a document containing multiple namespaces. 
> 
> The [`@namespace`](/en-US/docs/Web/CSS/@namespace) at-rule defined in this module enables distinguishing between same-named elements in different namespaces. Element tag names are not unique to a single language. For example, the `<a>` element isn't limited to HTML. You may want to style the `<a>`s  within your SVGs differently than links in your HTML. You also likely want to ensure that {{domxref("Document.querySelectorAll", "querySelectorAll(\"a\")")}} selects the right kind of element. Namespacing can help.
> 
> The `@namespace` rule is used for declaring a default namespace and for binding namespaces to namespace prefixes. The namespaces module also defines the syntax for using these prefixes to represent namespace-qualified names. That's it. What a name means or if the name is even valid depends on the context and host language.

Author: ```suggestion
CSS isn't just for styling HTML. A stylesheet may be used to style SVG, MathML, XML, or HTML, each of which has a different namespace or a document containing multiple namespaces.

The [`@namespace`](/en-US/docs/Web/CSS/@namespace) at-rule defined in this module enables distinguishing between same-named elements in different namespaces. Element tag names are not unique to a single language. For example, the `<a>` element isn't limited to HTML. You may want to style the `<a>`s within your SVGs differently than links in your HTML. You also likely want to ensure that {{domxref("Document.querySelectorAll", "querySelectorAll(\"a\")")}} selects the right kind of element. Namespacing can help.

The `@namespace` rule is used for declaring a default namespace and for binding namespaces to namespace prefixes. The namespaces module also defines the syntax for using these prefixes to represent namespace-qualified names. That's it. What a name means or if the name is even valid depends on the context and host language.
```
